packages/connect/src/core/index.ts

```diff
@@ -1149,3 +1162,311 @@ const disableWebUSBTransport = async () => {
+        }),
+    );
+    await waitForDisconnectedDevice(deviceList);
+
+    log.debug('onCallFirmwareUpdate', 'waiting for device to reconnect');
+    postMessage(
+        createUiMessage(UI.FIRMWARE_RECONNECT, {
+            device: device.toMessageObject(),
+            manual: flow === 'manual',
+            bootloader: true,
+            confirmOnDevice: flow !== 'manual',
+        }),
+    );
+    let reconnectedDevice = await waitForReconnectedDevice(deviceList, true);
+
+    await reconnectedDevice.acquire();
+
+    let stripped: ArrayBuffer | undefined;
+
+    while (reconnectedDevice.firmwareRelease?.isNewer) {
+        log.debug(
+            'onCallFirmwareUpdate',
+            'there is still a newer release available, proceed with install',
+        );
+        // each uploadFirmware should get a new binary. this is typically the case when intermediary is installed first
+        if (!binary) {
+            binary = await getBinaryHelper(reconnectedDevice);
+        }
+        // Might not be installed, but needed for calculateFirmwareHash anyway
+        stripped = stripFwHeaders(binary);
+        await uploadFirmware(
+            reconnectedDevice.getCommands().typedCall.bind(reconnectedDevice.getCommands()),
+            postMessage,
+            reconnectedDevice,
+            { payload: shouldStripFwHeaders(device.features) ? stripped : binary },
+        );
+        binary = undefined;
+
+        await reconnectedDevice.release();
+
+        log.debug('onCallFirmwareUpdate', 'waiting for device to disconnect');
+        postMessage(
+            createUiMessage(UI.FIRMWARE_DISCONNECT, {
+                device: device.toMessageObject(),
+                manual: flow === 'manual',
+            }),
+        );
+        await waitForDisconnectedDevice(deviceList);
+
+        log.debug('onCallFirmwareUpdate', 'waiting for device to reconnect in normal mode');
+        postMessage(
+            createUiMessage(UI.FIRMWARE_RECONNECT, {
+                device: device.toMessageObject(),
+                manual: flow === 'manual',
+                bootloader: false,
+                confirmOnDevice: flow !== 'manual',
+            }),
+        );
+        reconnectedDevice = await waitForReconnectedDevice(deviceList, false);
+        await reconnectedDevice.acquire();
+
+        // when custom binary is passed, don't try to updated to latest version
+        if (params.binary) {
+            break;
+        }
+    }
+
+    support.getFirmwareHash = reconnectedDevice.atLeast(['1.11.1', '2.5.1']) && !params.binary;
+
+    if (support.getFirmwareHash) {
+        log.debug('onCallFirmwareUpdate', 'getFirmwareHash supported, proceed with check');
+        const { hash, challenge } = calculateFirmwareHash(
+            device.features.major_version,
+            stripped!,
+            randomBytes(32),
+        );
+
+        try {
+            const getFirmwareHashResponse = await reconnectedDevice.commands!.typedCall(
```

Comment:
Would be nice to notify frontend about this similarly as installation progress is handled so that something can be displayed. We currently display a message that fw is being validated - otherwise there is quite a gap between 100% progress and being actually finished. Maybe firing an event when validation starts? But I guess I can also show the message when progress === 100 and the response has not been received yet.

Reply:
no problem, we could fire some more elaborate progress events. I was thinking about someting like

1. at the start of fwupdate process, fire an event with expected steps of fw update, so that you can display some empty bullet points or have idea about what 100% actually means.
2. fire those events as points from 1 are completed